hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerSelectionCriteria.java

```diff
@@ -88,19 +93,25 @@ private boolean isContainerReplicatingOrDeleting(ContainerID containerID) {
    * 5. Container should be closed.
    * 6. If the {@link LegacyReplicationManager} is enabled, then the container should not be an EC container.
    * @param node DatanodeDetails for which to find candidate containers.
-   * @return NavigableSet of candidate containers that satisfy the criteria.
+   * @return Set of candidate containers that satisfy the criteria.
    */
-  public NavigableSet<ContainerID> getCandidateContainers(
+  public Set<ContainerID> getCandidateContainers(
       DatanodeDetails node, long sizeMovedAlready) {
-    NavigableSet<ContainerID> containerIDSet =
-        new TreeSet<>(orderContainersByUsedBytes().reversed());
     try {
-      containerIDSet.addAll(nodeManager.getContainers(node));
+      // Initialize containerSet for node
+      if (!setMap.containsKey(node)) {
+        addNodeToSetMap(node);
+      }
+      // In case the node is removed
+      nodeManager.getContainers(node);
     } catch (NodeNotFoundException e) {
       LOG.warn("Could not find Datanode {} while selecting candidate " +
           "containers for Container Balancer.", node.toString(), e);
-      return containerIDSet;
+      setMap.remove(node);
+      return Collections.emptySet();
     }
+
+    NavigableSet<ContainerID> containerIDSet = setMap.get(node);
```

Comment:
A DN can have 1000s or more container sbased on size of disk at dn. So at scale, this needs further optimized.
we do not need filter all containers for the criteria of size and others, it can be filtered `dynamically on need basis`.

Reply:
I think we can keep the size ordering here, as mentioned above, clients most probably want a faster balancer then.

Reply:
I mean here, below logic is called for every container to verify if sizeMoved met and other condition
containerIDSet.removeIf(
        containerID -> shouldBeExcluded(containerID, node, sizeMovedAlready));

This check can be on need basis to avoid check over all containers till needed, like,
getNextMatchedContainer();

Reply:
Since the remove policy changes every time after generating each pair for move, I think it should be ok to do this filter every time generating candidate containers?

Reply:
The new approach is more capable of handling change in policy,

Reply:
Updated, PTAL.